Automattic/pocket-casts-ios
# [Watch] Only fetch the episode UUIDs upfront when rendering the UpNext list, lazy load the episodes as needed

Follow-up to #950, see that PR for more backstory.

On the Up Next table in the Apple Watch, instead of fetching all the episode objects from the database at once, only the UUIDs are fetched (which is an order of magnitude faster), and each episode object is lazy-loaded as needed. Loading a handful of episode objects is still fast enough so that scrolling through that list is smooth.

I focused on touching existing code as little as possible, and keeping overall changes to a minimum, since none of this is covered by tests.

I'm not happy with the following:
- The `isStub` property in the `BaseEpisode` class is a hack. I would rather create a `StubEpisode` class that implements the `BaseEpisode` protocol, but that would require tons of boilerplate.
- As requested by the feedback [here](https://github.com/Automattic/pocket-casts-ios/pull/950#issuecomment-1636192131), I created new methods on the storage layer instead of adding optional parameters to the existing ones. But naming is hard. I followed the naming pattern `getSomeEpisodes() -> getSomeEpisodeUuids()`, but in fact those new methods return `BaseEpisode` "dehydrated/anemic" objects, *not* UUID strings. Naming is hard. I considered returning just UUID strings, but that would just make the changes on the `EpisodeViewModel` classes more complicated.

Suggestions are welcome. There are no perfect solutions here, I'm just hoping the trade-offs I made align with what you consider acceptable.

## To test

- Have a big Up Next list. Adding "WTF With Marc Marron" for example should get you there, it's nearly 1000 episodes.
- On your Apple Watch, do stuff on the Up Next view: open it, pick an episode and move it up top, pick an episode and start playing it, etc.
- On trunk, any of those operations would result in a ~1s freeze (that has already improved a lot due to my previous PRs). With this PR, it should be near-instant.

I haven't included any tracing because this change will be immediately noticeable on a real Apple Watch. When I measured in Instruments with my ~800 ep library, I could barely even find a CPU spike when I opened the Up Next view.

## Checklist
